### ISSUE TITLE
Integrate with Jetbrains Compose

### ISSUE THREAD

**Issue by brianguertin** [msink/kotlin-libui#40]
I noticed in this project's README that you weren't sure about the DSL.

Well, what if we changed the DSL over to Android/Jetbrains Compose? It seems like fun idea because:

- Take advantage of the tree-building and reactivity built into compose
- Compose is gaining traction & mind-share, so this project might be more approachable by sharing concepts
- Improved code sharing; e.g. you could use `expect/actual` to build common interfaces that work across both Android and native desktop.
- Jetbrains is already working on Compose Desktop, but some devs (myself included) disapprove of using Skia canvas instead of native platform widgets for reasons of accessibility/etc.. Plus, it's on the JVM and not Kotlin/native. So we could create a nice alternative to fill a different niche.

I'm just brainstorming, but I'm hoping to have some time to explore this myself, thought the idea was worth sharing.

**Comment by brianguertin** [comment]
Yeah, well Jetbrains just announced their Kotlin/JS Compose, so they are putting the work in to support multiplatform. I'm hoping to see Jetbrains start an iOS port too, because that would be awesome, and would bring first-class support for Compose on Kotlin/Native. Maybe some day :crossed_fingers:

**Comment by brianguertin** [comment]
Looks like as of May 28th this is technically possible: https://twitter.com/shikasd_/status/1398445030419877891?s=20

Though, there are no prebuilt artifacts published for native (yet?)

**Comment by songdehuai** [comment]
Is it possible to use  LVLG .https://github.com/lvgl/lvgl

**Comment by msink** [comment]
LVGL is somewhat similar to compose-ui, yes - both implement all widgets by drawing them on plain canvas (framebuffer).
libui uses widgets from native os API

Totally different philosophy:
- LVGL/compose-ui look and feel is same on all systems, but differs from other programs on that system.
- libui on contrary - look different on every system - but similar to other programs.

If this progect ever will use compose (still possible, but unsure) - it will be lowlevel compose runtime, and maybe layout framework.
But not widgets. Drawing widgets on canvas is out of scope for this project.

**Comment by e13mort** [comment]
I think Compose is a great way to move. 

- It's designed to separate UI draw logic from runtime & layout logic. https://github.com/JakeWharton/mosaic is a good example of that concept. Compose for HTML is another.
- Compose Runtime has good interoperability with Kotlin Coroutines. Supporting it will allow one to create fully multiplatform applications natively for all major OSs. KMP + Compose UI for Android, KMP + Swift UI for iOS, KMP + `Composed libui` for Mac, Win & Linux

**Comment by Dominaezzz** [comment]
Seeing your comment caught my attention and I've started throwing together a POC/MVP.
I've managed to start up a window with a single label with compose!
I'll quickly port over ktx widgets and make a draft PR in a couple days.
It won't be tidy enough to merge as is but at least the bulk of the hard work would be done.

**Comment by Dominaezzz** [comment]
Made a PR. Take a look and give it a go. Lemme know if you encounter any bugs/crashes.
